pkg/kubelet/images/puller.go

```diff
@@ -40,22 +42,35 @@ type imagePuller interface {
 var _, _ imagePuller = &parallelImagePuller{}, &serialImagePuller{}
 
 type parallelImagePuller struct {
-	imageService kubecontainer.ImageService
-	tokens       chan struct{}
+	imageService    kubecontainer.ImageService
+	tokens          chan struct{}
+	pullingImageSet sync.Map
 }
 
 func newParallelImagePuller(imageService kubecontainer.ImageService, maxParallelImagePulls *int32) imagePuller {
 	if maxParallelImagePulls == nil || *maxParallelImagePulls < 1 {
-		return &parallelImagePuller{imageService, nil}
+		return &parallelImagePuller{imageService, nil, sync.Map{}}
+	}
+	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls), sync.Map{}}
+}
+
+func genKeyByImageSpec(spec kubecontainer.ImageSpec) string {
+	args := []string{spec.Image}
+	if spec.RuntimeHandler != "" {
```

Comment:
Read from [KEP-4216#L268-L271](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4216-image-pull-per-runtime-class/README.md?plain=1#L268-L271) 
>ImageSpec Annotations can be used to help the runtime target specific images in multi-arch images.
Should we decide on a new name like `runtime.kubernetes.io/architecture` different from ImageSpec.RuntimeHandler? ☺ @HirazawaUi

Reply:
These annotations are not part of Kubernetes' records. Perhaps a better approach for us would be to exclude certain annotations, such as `kubernetes.io/config.seen` and `kubectl.kubernetes.io/last-applied-configuration`, because the unique key they form will always make the image unique. However, I would prefer to hear the opinion of a sig-node reviewer."

Reply:
everything we pass as an argument to pullImage needs to be considered. We should not hardcode any assumptions about container runtime behavior into the kubelet

Reply:
Because of a large number of irrelevant annotations will always make the image unique, like @HirazawaUi mentioned. If we Include all as identify for image makes change meaningless.
I agree with Not hardcode any assumptions about container runtime behavior. 

So I have two suggestions ☺：
- **A:** Can we converge all contexts for CRI or image pull behavior to a unified annotation prefix? like `node.kubernetes.io/.*` ,`runtime.kubernetes.io/.*` ? 
- **B:** Or we define a "blacklist" annotation such as : `"node.kubernetes.io/runtime-no-effect":["kubernetes.io/config.seen","kubectl.kubernetes.io/last-applied-configuration","kubernetes.io/limit-ranger"]` If user needs to mask annotation which not related to runtime . We can filter them out in the beginning of [ImageSpec was defined](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/images/image_manager.go#L113C2-L119C3).

What do you think?  @SergeyKanzhelev @HirazawaUi

Reply:
cc: @mikebrow @samuelkarp @haircommander 

These two proposals generally describe two approaches - container runtime owning configuration on which labels it needs or kubelet ownes labels it thinks would be useful.

I would go as far as allowing CRI API to return the list of annotations (wildecards) it may accept on initialization. So container runtime can tell which labels it needs.

Reply:
"Allowing CRI API to return the list of annotations it may accept on initialization" seems to be a longer term feature. And I changed commit which solution ref to [126382#discussion](https://github.com/kubernetes/kubernetes/pull/126382#discussion_r1712158936)
Pull requests with the same image ID will be executed serially. It will resolve waiting "almost same" image pulls as well.

Reply:
> Pull requests with the same image ID will be executed serially. It will resolve waiting "almost same" image pulls as well.

Again, I feel that this assumption is too big. The last iteration is not accounting even for a handler.